lib/utils/updateCompiler.js

```diff
@@ -1,64 +1,43 @@
 'use strict';
 
 const webpack = require('webpack');
-const addEntries = require('./addEntries');
+const DevServerPlugin = require('./addEntries');
 const getSocketClientPath = require('./getSocketClientPath');
 
 function updateCompiler(compiler, options) {
-  const findHMRPlugin = (config) => {
-    if (!config.plugins) {
-      // eslint-disable-next-line no-undefined
-      return undefined;
-    }
+  const findPlugin = (config, plugin) =>
+    config.plugins.find((p) => p.constructor === plugin);
 
-    return config.plugins.find(
-      (plugin) => plugin.constructor === webpack.HotModuleReplacementPlugin
-    );
-  };
+  const compilers = compiler.compilers || [compiler];
+  // eslint-disable-next-line no-shadow
+  compilers.forEach((compiler) => {
+    const config = compiler.options;
+    config.plugins = config.plugins || [];
 
-  const compilers = [];
-  const compilersWithoutHMR = [];
-  if (compiler.compilers) {
-    // eslint-disable-next-line no-shadow
-    compiler.compilers.forEach((compiler) => {
-      compilers.push(compiler);
-      if (!findHMRPlugin(compiler.options)) {
-        compilersWithoutHMR.push(compiler);
-      }
-    });
-  } else {
-    compilers.push(compiler);
-    if (!findHMRPlugin(compiler.options)) {
-      compilersWithoutHMR.push(compiler);
+    let devServerPlugin = findPlugin(config, DevServerPlugin);
+    if (devServerPlugin) {
+      // use existing plugin
+      devServerPlugin.options = options;
+    } else {
+      devServerPlugin = new DevServerPlugin(options);
+      config.plugins.push(devServerPlugin);
     }
-  }
+    devServerPlugin.apply(compiler);
 
-  // it's possible that we should clone the config before doing
-  // this, but it seems safe not to since it actually reflects
-  // the changes we are making to the compiler
-  // important: this relies on the fact that addEntries now
-  // prevents duplicate new entries.
-  addEntries(compiler, options);
-  // eslint-disable-next-line no-shadow
-  compilers.forEach((compiler) => {
     const providePlugin = new webpack.ProvidePlugin({
       __webpack_dev_server_client__: getSocketClientPath(options),
     });
     providePlugin.apply(compiler);
-  });
 
-  // do not apply the plugin unless it didn't exist before.
-  if (options.hot === true || options.hot === 'only') {
-    // eslint-disable-next-line no-shadow
-    compilersWithoutHMR.forEach((compiler) => {
-      // addDevServerEntrypoints above should have added the plugin
-      // to the compiler options
-      const plugin = findHMRPlugin(compiler.options);
-      if (plugin) {
+    if (!findPlugin(config, webpack.HotModuleReplacementPlugin)) {
+      // add and apply the HMR plugin, if it didn't exist before.
+      if (options.hot === true || options.hot === 'only') {
+        const plugin = new webpack.HotModuleReplacementPlugin();
+        config.plugins.push(plugin);
```

Comment:
Why do not put this logic in the plugin too? Plugin can add other plugins :smile: It is normal

Reply:
I thought this is consistent, but since `DevServerPlugin` adds the hot entry it may make more sense to have it there. What do you think? Should I also move this too:
```
    const providePlugin = new webpack.ProvidePlugin({
      __webpack_dev_server_client__: getSocketClientPath(options),
    });
    providePlugin.apply(compiler);
```

Reply:
Yep, I think it make sense, ideally will be keeping all logic for modify compiler (add entries/plugins/change options and etc) should be happens in the plugin